### DISCUSSION TITLE
[.NET 7] Dynamically loading assemblies/WASM modules/...

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#10564] [Q&A] [answered]
@jeromelaban I'm looking (again) into opening up my app by dynamically loading assemblies/code at runtime to add features and to customize the UI.

**WindowsAppSDK**
Assemblies could be loaded e.g. by Assembly.Load(). It's my understanding that the loaded assembly could hold 'any' code including XAML. It would be great if you could let me know if your understanding is different. I know this is not UNO's court...

**UNO WASM**
A couple of questions come to mind:
- is dynamically loading WASM modules at runtime even supported by the way how UNO manages WASM? Would there by any guidance avalilable on how to do that?
- I take it such a UNO library - which is built for WindowsAppSDK and UNO WASM - would be created as per [here](https://platform.uno/docs/articles/get-started-dotnet-new.html?tabs=windows#uno-platform-blank-cross-runtime-library). Is that correct?
- would it simplify the tech stack if the developer of the custom library would not use XAML but use C# markup instead and e.g. avoid using XAML code generators (just speculating here) or does it basically not make much difference?

Lastly: are you aware of any road-blocks/show stoppers I might encounter when using latest UNO + MS bits to explore this avenue?

(I'm leaving aside the other heads for now, since they're not in my focus at this point. I know that dynamically loading code in iOS/MacOS would be no-go though)

Thanks in advance

**Comment by jeromelaban** [comment]
Uno uses the .NET runtime, which provides `Assembly.Load` APIs for loading assemblies, which you can use. The best way to know if this works for you is to try it out.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban. To clarify: you mean this even might be able to load WASM modules?

**Reply by jeromelaban** [reply]
Wasm modules are not supported dynamically at this time, only .NET assemblies.

**Reply by SongOfYouth** [reply]
@jeromelaban so, how to add the assemblies after build and find it at runtime?

**Comment by DierkDroth** [comment, marked as answer]
FYI
@jeromelaban and anyone who's working on that task

Turns out this is actually working - to my utmost surprise. 
- WindowsAppSDK: loading a custom/plugin assembly - even holding UI code - works as expected -> no problem
- UNO WASM, **no IL linker**: loading a custom/plugin assembly - even holding UI code - works as expected. Even support for different languages (RESX management) works no problem
- UNO WASM **+ IL linker**: here things are getting tricky - but ultimately do work. The problem in principle is that the IL linker [trims away stuff](https://platform.uno/docs/articles/features/using-il-linker-webassembly.html) which is needed

There are multiple steps to address that challenge:
1) Right on `Assembly.Load(From)` I saw an exception indicating that `System.Runtime` would not be found. This is addressed by adding `<assembly fullname="System.Runtime" />` to `LinkerConfig.xml`
2) Next I saw exceptions in browser console indicating that some attributes could not be constructed. I tried to address this by adding `System.Component.DataAnnotations` etc. (as per above) to `LinkerConfig.xml`. Unfortunately this did not work. Then I coded a 'dummy' method in my app which instantiated the offending attributes.  This tricks the IL linker and prevents trimming away code which is needed. 
Right now this method looks like this:
```
		public static object ILLinkerWorkAround()
		 => new object[]
					{
						new System.ComponentModel.BindableAttribute(true),
						new System.ComponentModel.EditorBrowsableAttribute(),
						new System.Diagnostics.DebuggerStepThroughAttribute(),
						new System.Runtime.CompilerServices.AsyncStateMachineAttribute(null),
					};
```
3) Then I still saw mysterious failures in browser console. After some tinkering I found that still assemblies would be trimmed away which are actually needed.
(BTW: `AppDomain.CurrentDomain.AssemblyLoad += (s, e) => Console.Out.WriteLine(e.LoadedAssembly.GetName().Name);` helps understanding which assemblies are loaded (IL linker on/off) to get a better understanding on what might be missing).
Ultimately I had to add 'printf' to isolate the failing code + adding exception handlers (global exception handling still does not appear to work) to understand what other assemblies might be missing. I ended up with this LinkerConfig.xml:
```
<linker>
	<assembly fullname="my-app-assembly1" />
	<assembly fullname="my-app-assembly2" />

	<assembly fullname="System.Core">
		<!-- This is required by JSon.NET and any expression.Compile caller -->
		<type fullname="System.Linq.Expressions*" />
	</assembly>

	<assembly fullname="Microsoft.Extensions.Options" />
	<assembly fullname="System.Runtime" />
	<assembly fullname="System.Threading" />
	<assembly fullname="Uno.UI" />
</linker>
```

Then - finally - the custom assemblies loaded by `Assembly.LoadFrom()` even as IL linker was active.

Lastly I tested if compiling the custom assembly to AOT would work and still the assembly could be loaded. Turns out: it works!
Here is the relevant section of my custom assembly UNO WASM project:
```
	<PropertyGroup Condition="'$(Configuration)'=='Debug'">
		<DefineConstants>$(DefineConstants);TRACE;DEBUG</DefineConstants>
		<DebugType>portable</DebugType>
		<DebugSymbols>true</DebugSymbols>
		<MonoRuntimeDebuggerEnabled>true</MonoRuntimeDebuggerEnabled>
		<WasmShellILLinkerEnabled>false</WasmShellILLinkerEnabled>
	</PropertyGroup>

	<PropertyGroup Condition="'$(Configuration)'=='Release'">
		<UnoXamlResourcesTrimming>true</UnoXamlResourcesTrimming>
		<WasmShellMonoRuntimeExecutionMode>InterpreterAndAOT</WasmShellMonoRuntimeExecutionMode>
	</PropertyGroup>
```

Obviously loading assemblies dynamically with IL linker on still is highly dependent on what features actually might be implemented in the custom assembly (and what assemblies might be needed). However, this is a 'learn as I go' issue...

**Comment by SongOfYouth** [comment]
@DierkDroth which is the right path to place the assembly in after build? 🌞

**Reply by DierkDroth** [reply]
@SongOfYouth I'm not exactly sure I understand your question. Ultimately you would have to place the assembly where `Assembly.LoadFrom()` would find it. Unfortunately, I don't recall where I placed it in my scenario as I ran my tests last year...

**Reply by SongOfYouth** [reply]
emm, thanks anyway.
